File: Data Structures & Algorithms/house-robber/submission-2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        '''
            f(0) = nums[0]
            f(1) = nums[1]
            f(i) = f(i-2) + nums[i]
                 = f(i-1)
        '''

        cache = [0] * len(nums)
        if len(nums) <= 2: return max(max(nums), 0)

        cache[0] = nums[0]
        cache[1] = max(nums[0], nums[1])

        for i in range(2, len(nums)):
            cache[i] = max(cache[i-1], cache[i-2] + nums[i])
            print(cache[i])
        print(cache)
        return cache[len(nums)-1]

File: Data Structures & Algorithms/maximum-product-subarray/submission-2.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size();
        vector<int> maxList(n, 1), minList(n, 1);
        maxList[0] = nums[0];
        minList[0] = nums[0];
        int val = nums[0];
        for(int i = 1; i < n; i++) {
            maxList[i] = max(nums[i], max(maxList[i-1] * nums[i], minList[i-1] * nums[i]));
            minList[i] = min(nums[i], min(maxList[i-1] * nums[i], minList[i-1] * nums[i]));
            val = max(val, maxList[i]);
        }
        return val;
    }
};
